### ISSUE TITLE
Open Waters: Easy onboarding for data scientists

### ISSUE THREAD

**Issue by santteegt** [oceanprotocol/openwaters#5]
# Open Waters : Easy onboarding for data scientists through a decentralized relayed network

## Abstract (required)

Currently, the onboarding process to Ocean features to data scientists provides a lot of friction, specifically when new users neither have enough background knowledge about how blockchain works nor experience executing transactions through decentralized applications: *DApp users usually need to jump through hoops — go to an exchange, pass a KYC process and wait for approval, transfer money from their bank, buy ETH, install MetaMask, learn how to use it and transfer ETH to it, etc*.

For interacting with a data marketplace or the squid library, besides owning some OCEAN tokens for staking/buying data assets, it is required to buy/own some ETH (or other native currency adopted in the [Ocean PoA network](https://blog.oceanprotocol.com/ocean-on-poa-vs-ethereum-mainnet-decd0ac72c97)) to pay for the transaction gas costs, which not only excludes the unbanked population, but also small organizations or individual data scientists looking to experiment with Ocean Protocol. Latest developments on meta transactions and decentralized relay protocols have allowed to ease the user onboading process on different dApps, which enable new alternatives for Ocean Protocol to accelerate user adoption for the new data economy.

## Motivation 

Extract of interactions within Gitter channel:

> Q
> Then, for the live network, would be great to integrate some meta transactions so the onboarding for  new users (even new to the concept of blockchain) can be done easily. I have worked with metaTx on my final project for the Consensys Dev Program, so I'm kind of familiar with this concept. Do you think it would be good to send this as a Open Waters bounty proposal?
>
> A
> To be honest, I don’t know of metaTX but, briefly google-ing it, it sounds super interesting! Do list out your thoughts as an open water bounty!
> one of our key outreach efforts is for data scientists. We are thinking of easy ways of how to on-board datascientists as, for a lot of them, Ethereum + the whole blockchain concept is quite new. If you can think of easy/interesting ways of onboarding this target audience, we can also collaborate on this

The main motivations of this proposal are:

* Trustless decentralized way of onboarding new users that don’t have ETH (or native network token) to pay for gas fees, and without requiring any changes to how Ocean Protocol works.
* Onboarding users quickly and effortlessly helps data marketplaces scale quickly. It’ll be easier to bootstrap dApp network effect if data providers/marketplaces subsidized some interactions with the keeper contract.

## Specification

Requirements to implement this OEP are:

* Implement a simpler, more decentralized and robust approach that provides a general purpose toll free interface for smart contracts
* Certain interactions with Keeper contracts should be subsidized by keeper nodes and/or data providers/marketplaces by allowing users to only sign a transaction with a key that doesn’t own ETH, and then wrapping it with another transaction signed by a key (relayer) that does own ETH, to pay for the gas
* To stay decentralized, all components have to be incentivized. Therefore, kepper nodes, data providers/marketplaces should be incentived for joining the pool of relayers for Ocean
* For onboarding use cases, instead of the public pool of transactions, we can use a public pool of relayers and let the sender select one.

## Implementation

This kind of solutions have already been implemented on various dApp solutions on the Ethereum network. An [article](https://medium.com/tabookey/1-800-ethereum-gas-stations-network-for-toll-free-transactions-4bbfc03a0a56) published by Tabookey introduced the definition and implementation of a `RelayHub` or decentralized relay network to ease the user onboarding process to dApps. Following, there is an extract of how it works:

### Gas stations

The network consists of gas stations (relays), publishing their fees and URLs on the blockchain, accepting etherless transactions from users, filling them with gas by signing them, returning them to the user, relaying them to the blockchain and getting compensated for their effort. Gas stations and users don’t trust each other, which leads us to…

### Anchor of Trust

A single open-source smart contract, RelayHub, acts as registrar for relays, mediates transactions between relays and dapp contracts, and emits events that help users find relays and assess their reputation. RelayHub holds small deposits from dapps and uses them to compensate relays for transactions accepted by dapps.

Relays are required to stake ETH in RelayHub when they register. The stake is used to prevent certain DoS or censorship attacks. Honest relays earn transaction fees, and provably dishonest relays lose their stake.

The entire interaction between a user and a relay is a short-lived http session. At the end of this http transaction the user gets a performance bond that keeps the relay honest, so the user knows that the transaction has been relayed without having to wait for mining to occur. A failed request can be detected immediately, and the user proceeds to the next relay. The network is censorship resistant as long as there is at least one honest relay.

### The user app

From the user’s perspective, the app (web or mobile) transparently selects a relay and delivers the transaction to the blockchain. Behind the scenes, the app selects a relay using on-chain information about transaction fees and ongoing reputation, attempts to send the transaction through the selected relay, knows immediately whether the request was handled, or retries with the next relay. The immediate validation process enables fast iteration, so malfunctioning or malicious relays cannot cause long delays, and the transaction gets delivered in a timely manner.

The full technical details, attacks and mitigations, are already available in [EIP 1613](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1613.md) draft.

Required smart contract components (RelayHub, RelayServer, RelayClient) can be inherited from this [example](https://github.com/tabookey/tabookey-gasless)

## References

* [1-800-ethereum-gas-stations-network-for-toll-free-transactions](https://medium.com/tabookey/1-800-ethereum-gas-stations-network-for-toll-free-transactions-4bbfc03a0a56)
* [Implementation example](https://medium.com/tabookey/how-to-make-your-dapp-gasless-da15d4d3cacd)
* [EIP-1613 draft](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1613.md)

## Open Questions

* **Which entities within Ocean Protocol can be a relayer?**: keeper nodes, data provides running Brizo, data marketplaces?
* **Incentive model**: how relayers should be rewarded? OCEAN tokens owned by the nework, or transaction senders?. User Adoption an "data liquidity" as incentive for data marketplaces to act as a relayers?

According to this [article](https://medium.com/tabookey/how-to-make-your-dapp-gasless-da15d4d3cacd), there are 4 steps to setting up a dapp to support gasless transactions (here some specific design decision should be taken into account within Ocean Protocol):

* **Decide on your access model**: Since you (the relay contract owner) pay for the gas, and the clients access it for free, you should decide which clients and which calls are allowed to access the contract.
In our example, we simply allow any client, but we will also show how you can charge your clients by other means (e.g. charge them with OCEAN tokens!).
* **Modify the contract to accept relayed calls**: Note that the contract always accepts “normal” call. We only add a mechanism to know who the relay caller is, when the call comes through a relay.
* **Modify your client**: make your web application (or mobile app) go through the relay to make calls.
Fill up your contract’s “gas tank” for the first time. Since the contract pays for the calls, you (as the contract owner) have to fund it periodically so it can pay the relay-servers for their service.

## Notes

The [EIP-1613 draft](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1613.md) provides a full description on how to implement a decentralized relay network

**Comment by ttmc** [comment]
I believe the current thinking is that anyone will be able to ask an Ocean PoA faucet for some Ocean PoA Ether, i.e. it will be free. The [Ocean Faucet Server](https://github.com/oceanprotocol/faucet) is already implemented.

**Comment by santteegt** [comment]
Yep, I've collaborated in the implementation of the faucet server, but faucets are useful for testnets. My proposal intends to ease the user onboarding on the mainnet through (subsidized) gassless interactions.

**Comment by ttmc** [comment]
The Ocean (Ether) faucet server can run in the Ocean testnets but I believe the current thinking is that _it will also be running on the Ocean Mainnet_ (also called the Pacific PoA Network, the Ocean PoA Mainnet and other things). It will give Ocean Mainnet Ether to whoever asks for it (for free).
